# Sequence-Alignment

## TODO
- [x] Make Global Alignment
- [x] Make Local Alignment
- [ ] Make Multiple Alignment
    - [x] Asses the produced MSA using three metrics
        - [x] Sum of Pairs
        - [x] Percentage of Identical Pairs *Suggestion idk*
        - [ ] Percentage of Conserved Pairs *Suggestion idk*
        - [x] Mutual Information
        - [ ] Normalized Mutual
    - [x] Plot the results
        - [x] Plot the Phylogenetic Tree
        - [x] Plot the Sequence logo
        - [x] Plot the Dot Plot
- [x] Make a GUI
    - [x] Add Input field and drag and drop field and browse button
    - [x] View the alignment with the alignment score
    - [x] Add a button to save the alignment
    - [x] Create unique plots for the alignment results
- [x] Validate That the actual input is a fasta file
- [ ] Make a CLI
    - [x] Add a command line interface
    - [ ] Add a command to save the alignment

## How to run the program
You need to install ghostscript to be able to make Sequence Logo from here: https://ghostscript.com/releases/gsdnld.html


