# Sequence-Alignment

## TODO
- [x] Make Global Alignment
- [x] Make Local Alignment
- [ ] Make Multiple Alignment
    - [ ] Asses the produced MSA using three metrics
        - [ ] Sum of Pairs
        - [ ] Percentage of Identical Pairs *Suggestion idk*
        - [ ] Percentage of Conserved Pairs *Suggestion idk*
        - [ ] Mutual Information
        - [ ] Normalized Mutual
    - [ ] Plot the results
        - [x] Plot the Phylogenetic Tree
        - [ ] Plot the Sequence logo
        - [x] Plot the Dot Plot
- [ ] Make a GUI
    - [x] Add Input field and drag and drop field and browse button
    - [x] View the alignment with the alignment score
    - [x] Add a button to save the alignment
    - [ ] Create unique plots for the alignment results
- [ ] Validate That the actual input is a fasta file
- [ ] Make a CLI
    - [x] Add a command line interface
    - [ ] Add a command to save the alignment

